plugins/bundle/plugin.go

```diff
@@ -14,34 +14,44 @@ import (
 
 	"github.com/open-policy-agent/opa/ast"
 	"github.com/open-policy-agent/opa/bundle"
+	"github.com/open-policy-agent/opa/bundle/manifest"
 	"github.com/open-policy-agent/opa/download"
-	"github.com/open-policy-agent/opa/internal/manifest"
 	"github.com/open-policy-agent/opa/plugins"
 	"github.com/open-policy-agent/opa/storage"
 	"github.com/sirupsen/logrus"
 )
 
 // Plugin implements bundle activation.
 type Plugin struct {
-	config     Config
-	manager    *plugins.Manager             // plugin manager for storage and service clients
-	status     *Status                      // current plugin status
-	etag       string                       // etag on last successful activation
-	listeners  map[interface{}]func(Status) // listeners to send status updates to
-	downloader *download.Downloader
-	mtx        sync.Mutex
+	config        Config
+	manager       *plugins.Manager                         // plugin manager for storage and service clients
+	status        map[string]*Status                       // current status for each bundle
+	etags         map[string]string                        // etag on last successful activation
+	listeners     map[interface{}]func(Status)             // listeners to send status updates to
+	bulkListeners map[interface{}]func(map[string]*Status) // listeners to send aggregated status updates to
+	downloaders   map[string]*download.Downloader
+	mtx           sync.Mutex
+	cfgMtx        sync.Mutex
```

Comment:
Can you add a comment about why cfgMtx is required. It looks like it's just so that concurrent calls to Reconfigure are safe.

Reply:
Yep, comment added in latest version. You're right that it is just for the case of reconfigure in parallel (it would very likely cause problems). It used to just grab the `p.mtx` but if we do that and stop a downloader you can get a deadlock with the downloader processing an update (it tries to call the plugin oneshot, but gets stuck waiting on the `p.mtx` which is waiting for the stop..). Anyway, this keeps the protection on reconfigure and removes that issue.